client/src/commands/getTcb.ts

```diff
@@ -0,0 +1,120 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import * as vscode from 'vscode';
+
+import * as lspext from '../../common/lsp_extensions';
+import {AngularLanguageClient} from '../client';
+
+const TCB_HIGHLIGHT_DECORATION = vscode.window.createTextEditorDecorationType({
+  backgroundColor: new vscode.ThemeColor('editor.selectionHighlightBackground'),
+});
+
+class SourceMap {
```

Comment:
I don't think we need this class. Can't we just convert `lsp.Range` to `vscode.Range` by calling the `protocol2CodeConverter`?

Reply:
Yes, but the problem is that we don't have `lsp.Range`. We have `ts.TextSpan`, which is a pair of character offsets (start, end) which we need to convert into line/column pairs.

Reply:
I'm doing a similar conversion in my work with "find components for template" and I'm now using `tsTextSpanToLspRange` in `session.ts` and then from there, converting it to `vscode.Range` in the command. I think that would do the trick here as well?

Reply:
We also cannot convert from TextSpan to Range on the server side because to do so we would have to assume that the TCB belongs to a script info; while this is true (IIRC), I don't think that's an assumption that the server should make with regard to how the language service handles TCBs and whether it does/doesn't attach them onto a project.

Reply:
> we would have to assume that the TCB belongs to a script info ...

This will always be the case. Since the server owns the `ProjectService`, which in turn manages all `ScriptInfo`s, converting TS spans to LSP Range should be done by the server. We do the conversion for other methods too, like definition, references, etc.
In fact, it is a leak of abstraction to send TS spans down to the client, because the client should not be aware of TS at all. It only deals with LSP.

Reply:
Would it not be a leaky abstraction to assume that the typecheck block file is attached to the project? This is an internal detail of how the language service creates TCB files and may change later.

I agree the conversion boundary should happen on the server side, you bring up a great point.